File: apps/mobile-poc/android/app/src/main/java/cn/edu/qlu/toolbox/mobilepoc/GradeExportCoordinator.java
package cn.edu.qlu.toolbox.mobilepoc;

import com.getcapacitor.JSObject;
import java.lang.ref.WeakReference;

final class GradeExportCoordinator {
    interface EventSink { void send(JSObject event); }

    private static EventSink eventSink;
    private static WeakReference<GradeExportActivity> activity = new WeakReference<>(null);

    private GradeExportCoordinator() {}

    static synchronized void setEventSink(EventSink sink) { eventSink = sink; }
    static synchronized void attach(GradeExportActivity value) { activity = new WeakReference<>(value); }
    static synchronized void detach(GradeExportActivity value) {
        if (activity.get() == value) activity.clear();
    }
    static synchronized boolean hasActiveTask() {
        GradeExportActivity value = activity.get();
        return value != null && value.isTaskActive();
    }
    static synchronized void emit(JSObject event) {
        if (eventSink != null) eventSink.send(event);
    }
    static synchronized boolean continueAfterLogin(String taskId) {
        GradeExportActivity value = activity.get();
        return value != null && value.continueAfterLogin(taskId);
    }
    static synchronized boolean cancel(String taskId) {
        GradeExportActivity value = activity.get();
        return value != null && value.cancelTask(taskId);
    }
}

File: apps/mobile/android/app/src/main/java/cn/edu/qlu/toolbox/GradeExportCoordinator.kt
package cn.edu.qlu.toolbox

import com.getcapacitor.JSObject
import java.lang.ref.WeakReference

internal object GradeExportCoordinator {
    private var activity = WeakReference<GradeExportActivity>(null)
    private var eventSink: ((JSObject) -> Unit)? = null
    private var expectedLaunchTaskId: String? = null

    @Synchronized fun expectLaunch(taskId: String) { expectedLaunchTaskId = taskId }
    @Synchronized fun consumeExpectedLaunch(taskId: String): Boolean {
        if (expectedLaunchTaskId != taskId) return false
        expectedLaunchTaskId = null
        return true
    }
    @Synchronized fun attach(value: GradeExportActivity) { activity = WeakReference(value) }
    @Synchronized fun detach(value: GradeExportActivity) { if (activity.get() === value) activity.clear() }
    @Synchronized fun setEventSink(sink: (JSObject) -> Unit) { eventSink = sink }
    @Synchronized fun emit(event: JSObject) { eventSink?.invoke(event) }
    @Synchronized fun hasAttachedActivity(): Boolean = activity.get()?.isTaskActive() == true
    @Synchronized fun cancel(taskId: String): Boolean = activity.get()?.cancelTask(taskId) == true
}
